wdl-analysis/src/queue.rs

```diff
@@ -21,146 +21,238 @@ use tokio::sync::oneshot;
+                    } else {
+                        log::info!("received request to analyze all documents");
+                    }
+
+                    let results = self.analyze(document, context, &completed);
+
                     log::info!(
-                        "received request to remove {count} URI(s)",
-                        count = uris.len()
+                        "request to analyze documents completed in {elapsed:?}",
+                        elapsed = start.elapsed()
                     );
 
-                    self.remove_documents(uris, completed);
+                    completed.send(results).ok();
+                }
+                Request::Remove(RemoveRequest {
+                    documents,
+                    completed,
+                }) => {
+                    let start = Instant::now();
+                    log::info!(
+                        "received request to remove {count} documents(s)",
+                        count = documents.len()
+                    );
+
+                    self.remove_documents(documents);
 
                     log::info!(
-                        "removal request completed in {elapsed:?}",
+                        "request to remove documents completed in {elapsed:?}",
                         elapsed = start.elapsed()
                     );
+
+                    completed.send(()).ok();
+                }
+                Request::NotifyIncrementalChange(NotifyIncrementalChangeRequest {
+                    document,
+                    change,
+                }) => {
+                    let mut graph = self.graph.write();
+                    if let Some(node) = graph.get_index(&document) {
+                        graph.get_mut(node).notify_incremental_change(change);
+                    }
+                }
+                Request::NotifyChange(NotifyChangeRequest {
+                    document,
+                    discard_pending,
+                }) => {
+                    let mut graph = self.graph.write();
+                    if let Some(node) = graph.get_index(&document) {
+                        graph.get_mut(node).notify_change(discard_pending);
+                    }
                 }
             }
         }
 
         log::info!("analysis queue has shut down");
     }
 
+    /// Adds a set of documents to the document graph.
+    fn add_documents(&self, documents: IndexSet<Url>) {
+        let mut graph = self.graph.write();
+        for document in documents {
+            graph.add_node(document, true);
+        }
+    }
+
     /// Analyzes the requested documents.
     fn analyze(
         &self,
-        documents: Vec<Arc<Url>>,
-        context: Option<String>,
-        completed: oneshot::Sender<Vec<AnalysisResult>>,
-    ) {
+        document: Option<Url>,
+        context: Context,
+        completed: &oneshot::Sender<Vec<AnalysisResult>>,
+    ) -> Vec<AnalysisResult> {
```

Comment:
Nit: since this is consumer facing, it should probably be an enum that tells the caller
what happened (`T::Cancelled`, `T::Completed(Vec<AnalysisResult)`, etc).